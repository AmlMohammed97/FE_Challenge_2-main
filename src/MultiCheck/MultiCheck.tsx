import CheckBox from './CheckBox';
import React from 'react';

export type Option = {
  label: string;
  value: string;
};

/**
 * Notice:
 * 1. There should be a special `Select All` option with checkbox to control all passing options
 * 2. If columns > 1, the options should be placed from top to bottom in each column
 *
 * @param {string} label - the label text of this component
 * @param {Option[]} options - options
 * @param {string[]} values - default checked option values
 * @param {number} columns - default value is 1
 * @param {Function} onChange - when checked options are changed,
 *                             they should be passed to outside
 */
type Props = {
  label?: string;
  options: Option[];
  columns?: number;
  values: string[];
  onChange: (values: string[]) => void;
};

const MultiCheck: React.FunctionComponent<Props> = ({ label, options, columns = 1, values, onChange }): JSX.Element => {
  const onChangeCheckbox = (optionValue: string) => {
    // if value selected, then un-select it
    if (values.includes(optionValue)) onChange(values.filter((value) => value !== optionValue));
    // otherwise select it
    else onChange([...values, optionValue]);
  };

  const isAllOptionsSelected = () => {
    let isAllSelected = true;
    options.map((option) => {
      // checked if and only if all other options are checked.
      // Unchecked if any other option is unchecked.
      if (!values.includes(option.value)) isAllSelected = false;
    });
    return isAllSelected;
  };

  const onClickSelectAll = () => {
    if (isAllOptionsSelected()) onChange([]); // Unchecked: All other options are unchecked.
    else onChange(options.map((option) => option.value)); // Checked: All other options are checked.
  };

  return (
    <div className="MultiCheck">
      <h1>{label}</h1>
      <div data-testid="MultiCheckColumns" style={{ columnCount: columns }}>
        <CheckBox label="Select All" dataTestId="SelectAll" key="SelectAll" onChange={onClickSelectAll} isChecked={isAllOptionsSelected()} />
        {options.map((option) => (
          <CheckBox label={option.label} dataTestId={option.value} key={option.value} onChange={() => onChangeCheckbox(option.value)} isChecked={values?.includes(option.value) ?? false} />
        ))}
      </div>
    </div>
  );
};

export default MultiCheck;
