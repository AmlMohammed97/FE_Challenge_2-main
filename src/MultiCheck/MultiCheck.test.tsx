import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MultiCheck, { Option } from '.';

const options: Option[] = [
  { label: 'aaa', value: '111' },
  { label: 'bbb', value: '222' },
  { label: 'ccc', value: '333' },
];

describe('MultiCheck', () => {
  describe('initialize', () => {
    it('renders the label if label provided', () => {
      render(<MultiCheck label="test checkbox label" onChange={() => {}} options={[]} values={[]} />);
      const label = screen.getByText('test checkbox label');
      expect(label).toBeInTheDocument();
    });
    it('renders with one column', () => {
      render(<MultiCheck label="" onChange={() => {}} options={[]} values={[]} />);
      const MultiCheckColumns = screen.getByTestId('MultiCheckColumns');
      expect(MultiCheckColumns).toHaveStyle({ 'column-count': '1' });
    });
    it('renders with three columns', () => {
      render(<MultiCheck label="" onChange={() => {}} options={[]} columns={3} values={[]} />);
      const MultiCheckColumns = screen.getByTestId('MultiCheckColumns');
      expect(MultiCheckColumns).toHaveStyle({ 'column-count': '3' });
    });
    it('renders unchecked select all checkbox', () => {
      render(<MultiCheck label="" onChange={() => {}} options={options} values={[]} />);
      const selectAllCheckbox = screen.getByTestId('SelectAll');
      expect(selectAllCheckbox).toBeInTheDocument();
      expect(selectAllCheckbox).not.toBeChecked();
    });
    it('renders checked select all checkbox', () => {
      render(<MultiCheck label="" onChange={() => {}} options={options} values={['111', '222', '333']} />);
      const selectAllCheckbox = screen.getByTestId('SelectAll');
      expect(selectAllCheckbox).toBeInTheDocument();
      expect(selectAllCheckbox).toBeChecked();
    });
    it('renders with two unchecked options', () => {
      render(<MultiCheck label="" onChange={() => {}} options={options} values={[]} />);
      const firstOption = screen.getByText('aaa');
      expect(firstOption).toBeInTheDocument();
      expect(firstOption).not.toBeChecked();
      const secondOption = screen.getByText('bbb');
      expect(secondOption).toBeInTheDocument();
      expect(secondOption).not.toBeChecked();
    });
  });
});
