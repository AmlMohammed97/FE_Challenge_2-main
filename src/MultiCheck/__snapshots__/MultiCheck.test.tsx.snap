// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCheck snapshot renders correctly 1`] = `
<div
  className="MultiCheck"
>
  <h1>
    test checkbox label
  </h1>
  <div
    data-testid="ListOfComponents"
    style={
      {
        "columnCount": 3,
      }
    }
  >
    <div>
      <input
        checked={true}
        data-testid="111"
        onChange={[Function]}
        type="checkbox"
      />
      <label>
        aaa
      </label>
    </div>
    <div>
      <input
        checked={true}
        data-testid="222"
        onChange={[Function]}
        type="checkbox"
      />
      <label>
        bbb
      </label>
    </div>
    <div>
      <input
        checked={true}
        data-testid="333"
        onChange={[Function]}
        type="checkbox"
      />
      <label>
        ccc
      </label>
    </div>
  </div>
</div>
`;
